torch/jit/__init__.py

```diff
@@ -155,9 +130,43 @@ def __call__(cls, *args, **kwargs):
             # user passed in), this problem goes away, because the class
             # __init__ is a part of is indeed Foo.
 
+            old_init = arg.__init__
+            # Python 2 has a concept of unbound methods, which doesn't play well
+            # with what we want to do here. Therefore, unpack the method into a
+            # regular function. Python 3 already returns a plain function, so
+            # nothing has to be done.
+            if sys.version_info[0] == 2:
+                old_init = old_init.im_func
+
+            def __init__(self, *args, **kwargs):
+                torch._C.CompiledFunction.__init__(self,
+                                                   nderivs, optimize, enabled,
+                                                   self.forward,
+                                                   lambda: list(self.parameters()),
+                                                   arg.__name__)
+                try:
+                    old_init(self, *args, **kwargs)
+                except TypeError as e:
+                    # If this fails here, the user probably didn't use this as a class decorator
+                    if "super" in str(e):
+                        raise_from(TypeError("torch.jit.compile must be used as a class decorator; "
+                                             "using it on an already defined class is not valid."
+                                             "\n\nOriginal error: {}".format(str(e))), e)
+                    else:
+                        raise
+                # NOTE: This can't be done in CompiledFunction constructor,
+                # because self.parameters() isn't well defined by then
+                # (Module constructor hasn't run yet).
+                self.update_captured_vars()
```

Comment:
Is it possible to make this just `set_captured_vars()`? The lambda callback plus this updater makes it tricky to understand what is happening. `self.set_captured_vars(list(self.parameters())` seems simpler.

Reply:
It is. I thought we'll have to call this from time to time (e.g. after each type cast/device change), but we capture references to Parameters, and these changes only affect their `.data`, so they don't get stale